File: opinionpulse/opinionpulse-backend/app/api/routes/projects.py
from fastapi import APIRouter, Depends, status
from sqlalchemy.orm import Session

from app.api.deps import get_current_user
from app.api.project_deps import get_owned_project
from app.db.database import get_db
from app.db.models import Project, Source, User
from app.schemas.project import (
    ProjectCreate,
    ProjectListResponse,
    ProjectResponse,
    ProjectUpdate,
)
from app.schemas.source import DEFAULT_SOURCES

router = APIRouter(prefix="/api/projects", tags=["projects"])


def _create_default_sources(project_id: int, db: Session) -> None:
    for item in DEFAULT_SOURCES:
        db.add(
            Source(
                project_id=project_id,
                source_name=item["source_name"],
                is_enabled=item["is_enabled"],
            )
        )


@router.get("", response_model=ProjectListResponse)
def list_projects(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    projects = (
        db.query(Project)
        .filter(Project.user_id == current_user.id)
        .order_by(Project.created_at.desc())
        .all()
    )
    return ProjectListResponse(
        projects=[ProjectResponse.model_validate(p) for p in projects],
        total=len(projects),
    )


@router.post("", response_model=ProjectResponse, status_code=status.HTTP_201_CREATED)
def create_project(
    payload: ProjectCreate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    project = Project(
        user_id=current_user.id,
        name=payload.name.strip(),
        description=payload.description,
        tracking_frequency=payload.tracking_frequency,
    )
    db.add(project)
    db.flush()
    _create_default_sources(project.id, db)
    db.commit()
    db.refresh(project)
    return ProjectResponse.model_validate(project)


@router.get("/{project_id}", response_model=ProjectResponse)
def get_project(
    project_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    project = get_owned_project(project_id, current_user, db)
    return ProjectResponse.model_validate(project)


@router.put("/{project_id}", response_model=ProjectResponse)
def update_project(
    project_id: int,
    payload: ProjectUpdate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    project = get_owned_project(project_id, current_user, db)
    update_data = payload.model_dump(exclude_unset=True)
    if "name" in update_data and update_data["name"]:
        update_data["name"] = update_data["name"].strip()
    for field, value in update_data.items():
        setattr(project, field, value)
    db.commit()
    db.refresh(project)
    return ProjectResponse.model_validate(project)


@router.delete("/{project_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_project(
    project_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    project = get_owned_project(project_id, current_user, db)
    db.delete(project)
    db.commit()
    return None

File: opinionpulse/opinionpulse-frontend/src/api/projects.ts
import { apiRequest } from "@/api/client"

export type Project = {
  id: number
  user_id: number
  name: string
  description: string | null
  tracking_frequency: string
  created_at: string
  updated_at: string
}

export type ProjectListResponse = {
  projects: Project[]
  total: number
}

export type ProjectCreateData = {
  name: string
  description?: string
  tracking_frequency: "manual" | "daily" | "weekly"
}

export type ProjectUpdateData = Partial<ProjectCreateData>

export function getProjects() {
  return apiRequest<ProjectListResponse>("/api/projects", { auth: true })
}

export function createProject(data: ProjectCreateData) {
  return apiRequest<Project>("/api/projects", {
    method: "POST",
    body: data,
    auth: true,
  })
}

export function getProject(id: number) {
  return apiRequest<Project>(`/api/projects/${id}`, { auth: true })
}

export function updateProject(id: number, data: ProjectUpdateData) {
  return apiRequest<Project>(`/api/projects/${id}`, {
    method: "PUT",
    body: data,
    auth: true,
  })
}

export function deleteProject(id: number) {
  return apiRequest<void>(`/api/projects/${id}`, {
    method: "DELETE",
    auth: true,
  })
}
